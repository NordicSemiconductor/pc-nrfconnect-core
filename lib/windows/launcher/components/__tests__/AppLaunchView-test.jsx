/* Copyright (c) 2015 - 2017, Nordic Semiconductor ASA
 *
 * All rights reserved.
 *
 * Use in source and binary forms, redistribution in binary form only, with
 * or without modification, are permitted provided that the following conditions
 * are met:
 *
 * 1. Redistributions in binary form, except as embedded into a Nordic
 *    Semiconductor ASA integrated circuit in a product or a software update for
 *    such product, must reproduce the above copyright notice, this list of
 *    conditions and the following disclaimer in the documentation and/or other
 *    materials provided with the distribution.
 *
 * 2. Neither the name of Nordic Semiconductor ASA nor the names of its
 *    contributors may be used to endorse or promote products derived from this
 *    software without specific prior written permission.
 *
 * 3. This software, with or without modification, must only be used with a Nordic
 *    Semiconductor ASA integrated circuit.
 *
 * 4. Any software provided in binary form under this license must not be reverse
 *    engineered, decompiled, modified and/or disassembled.
 *
 * THIS SOFTWARE IS PROVIDED BY NORDIC SEMICONDUCTOR ASA "AS IS" AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
 * MERCHANTABILITY, NONINFRINGEMENT, AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL NORDIC SEMICONDUCTOR ASA OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR
 * TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

import React from 'react';
import { List } from 'immutable';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import AppLaunchView from '../AppLaunchView';
import AppItemButton from '../AppItemButton';
import getImmutableApp from '../../models';

describe('AppLaunchView', () => {
    it('should render without any apps', () => {
        expect(renderer.create(
            <AppLaunchView
                apps={List()}
                onAppSelected={() => {}}
                onMount={() => {}}
                isRetrievingApps={false}
            />,
        )).toMatchSnapshot();
    });

    it('should render spinner while retrieving apps', () => {
        expect(renderer.create(
            <AppLaunchView
                apps={List()}
                onAppSelected={() => {}}
                onMount={() => {}}
                isRetrievingApps
            />,
        )).toMatchSnapshot();
    });

    it('should render apps sorted by name and isOfficial', () => {
        const templateApp = getImmutableApp({
            currentVersion: '1.2.3',
            latestVersion: '1.2.3',
            engineVersion: '2.x',
            isSupportedEngine: true,
        });
        expect(renderer.create(
            <AppLaunchView
                apps={
                    List([
                        templateApp
                            .set('name', 'appB')
                            .set('isOfficial', false)
                            .set('path', '/path/to/local/B'),
                        templateApp
                            .set('name', 'appB')
                            .set('isOfficial', true)
                            .set('path', '/path/to/B'),
                        templateApp
                            .set('name', 'appA')
                            .set('isOfficial', false)
                            .set('path', '/path/to/local/A'),
                        templateApp
                            .set('name', 'appA')
                            .set('isOfficial', true)
                            .set('path', '/path/to/A'),
                        templateApp
                            .set('name', 'appA')
                            .set('isOfficial', true)
                            .set('path', '/path/to/otherA'),
                    ])
                }
                onAppSelected={() => {}}
                onMount={() => {}}
                isRetrievingApps={false}
            />,
        )).toMatchSnapshot();
    });

    it('should render app with no engine version', () => {
        const app = getImmutableApp({
            name: 'pc-nrfconnect-foobar',
            currentVersion: '1.2.3',
            path: '/path/to/pc-nrfconnect-foobar',
            isOfficial: false,
        });
        expect(renderer.create(
            <AppLaunchView
                apps={List([app])}
                onMount={() => {}}
                onAppSelected={() => {}}
                isRetrievingApps={false}
            />,
        )).toMatchSnapshot();
    });

    it('should render app with unsupported engine version', () => {
        const app = getImmutableApp({
            name: 'pc-nrfconnect-foobar',
            currentVersion: '1.2.3',
            path: '/path/to/pc-nrfconnect-foobar',
            isOfficial: false,
            engineVersion: '1.x',
            isSupportedEngine: false,
        });
        expect(renderer.create(
            <AppLaunchView
                apps={List([app])}
                onMount={() => {}}
                onAppSelected={() => {}}
                isRetrievingApps={false}
            />,
        )).toMatchSnapshot();
    });

    it('should render app with upgrade available', () => {
        const app = getImmutableApp({
            name: 'pc-nrfconnect-foobar',
            currentVersion: '1.2.3',
            latestVersion: '1.2.4',
            path: '/path/to/pc-nrfconnect-foobar',
            isOfficial: true,
            engineVersion: '1.x',
            isSupportedEngine: true,
        });
        expect(renderer.create(
            <AppLaunchView
                apps={List([app])}
                onMount={() => {}}
                onAppSelected={() => {}}
                isRetrievingApps={false}
            />,
        )).toMatchSnapshot();
    });

    it('should invoke onAppSelected with given app item when app is clicked', () => {
        const app = getImmutableApp({
            name: 'pc-nrfconnect-foobar',
            currentVersion: '1.2.3',
            path: '/path/to/pc-nrfconnect-foobar',
            isOfficial: false,
        });
        const onAppSelected = jest.fn();
        const wrapper = mount(
            <AppLaunchView
                apps={List([app])}
                onAppSelected={onAppSelected}
                onMount={() => {}}
                isRetrievingApps={false}
            />,
        );
        wrapper.find(AppItemButton).first().simulate('click');

        expect(onAppSelected).toHaveBeenCalledWith(app);
    });
});
