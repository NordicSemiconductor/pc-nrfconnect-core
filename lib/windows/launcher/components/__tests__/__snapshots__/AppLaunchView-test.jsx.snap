// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppLaunchView should render app with no engine version 1`] = `
<div
  className="list-group"
>
  <div
    className="core-app-launch-item list-group-item"
  >
    <div
      className="core-app-icon"
    >
      <img
        alt="App icon"
        draggable={false}
        src="test-file-stub"
      />
      <div>
        <span
          className="alert-icon-bg"
        />
        <span
          className="glyphicon glyphicon-alert"
          title="The app does not specify which nRF Connect version(s) it supports"
        />
      </div>
    </div>
    <div>
      <h4
        className="list-group-item-heading"
      >
        pc-nrfconnect-foobar
      </h4>
      <p
        className="list-group-item-text"
      >
        local
        , v
        1.2.3
      </p>
    </div>
    <div
      className="core-app-launch-item-buttons btn-toolbar"
    >
      <button
        className="btn btn-primary core-btn"
        disabled={false}
        onClick={[Function]}
        title="Launch app"
        type="button"
      >
        <span
          className="glyphicon glyphicon-play"
        />
        <span
          className="core-btn-text"
        >
          Launch
        </span>
      </button>
      <div
        className="dropdown btn-group core-dropdown"
      >
        <div
          className="dropdown btn-group"
        >
          <button
            aria-expanded={false}
            aria-haspopup={true}
            className="btn btn-primary core-btn dropdown-toggle btn btn-default"
            disabled={false}
            id="bg-nested-dropdown"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            type="button"
          >
             
            <span
              className="caret"
            />
          </button>
          <ul
            aria-labelledby="bg-nested-dropdown"
            className="dropdown-menu dropdown-menu-right"
            role="menu"
          >
            <li
              className=""
              role="presentation"
            >
              <a
                href="#"
                onClick={[Function]}
                onKeyDown={[Function]}
                role="menuitem"
                tabIndex="-1"
              >
                <button
                  className="btn btn-primary core-btn"
                  disabled={false}
                  onClick={[Function]}
                  title="Create a desktop shortcut for this app"
                  type="button"
                >
                  <span
                    className="glyphicon glyphicon-share-alt"
                  />
                  <span
                    className="core-btn-text"
                  >
                    Create shortcut
                  </span>
                </button>
              </a>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`AppLaunchView should render app with unsupported engine version 1`] = `
<div
  className="list-group"
>
  <div
    className="core-app-launch-item list-group-item"
  >
    <div
      className="core-app-icon"
    >
      <img
        alt="App icon"
        draggable={false}
        src="test-file-stub"
      />
      <div>
        <span
          className="alert-icon-bg"
        />
        <span
          className="glyphicon glyphicon-alert"
          title="The app only supports nRF Connect 1.x, which does not match your currently installed version"
        />
      </div>
    </div>
    <div>
      <h4
        className="list-group-item-heading"
      >
        pc-nrfconnect-foobar
      </h4>
      <p
        className="list-group-item-text"
      >
        local
        , v
        1.2.3
      </p>
    </div>
    <div
      className="core-app-launch-item-buttons btn-toolbar"
    >
      <button
        className="btn btn-primary core-btn"
        disabled={false}
        onClick={[Function]}
        title="Launch app"
        type="button"
      >
        <span
          className="glyphicon glyphicon-play"
        />
        <span
          className="core-btn-text"
        >
          Launch
        </span>
      </button>
      <div
        className="dropdown btn-group core-dropdown"
      >
        <div
          className="dropdown btn-group"
        >
          <button
            aria-expanded={false}
            aria-haspopup={true}
            className="btn btn-primary core-btn dropdown-toggle btn btn-default"
            disabled={false}
            id="bg-nested-dropdown"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            type="button"
          >
             
            <span
              className="caret"
            />
          </button>
          <ul
            aria-labelledby="bg-nested-dropdown"
            className="dropdown-menu dropdown-menu-right"
            role="menu"
          >
            <li
              className=""
              role="presentation"
            >
              <a
                href="#"
                onClick={[Function]}
                onKeyDown={[Function]}
                role="menuitem"
                tabIndex="-1"
              >
                <button
                  className="btn btn-primary core-btn"
                  disabled={false}
                  onClick={[Function]}
                  title="Create a desktop shortcut for this app"
                  type="button"
                >
                  <span
                    className="glyphicon glyphicon-share-alt"
                  />
                  <span
                    className="core-btn-text"
                  >
                    Create shortcut
                  </span>
                </button>
              </a>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`AppLaunchView should render app with upgrade available 1`] = `
<div
  className="list-group"
>
  <div
    className="core-app-launch-item list-group-item"
  >
    <div
      className="core-app-icon"
    >
      <img
        alt="App icon"
        draggable={false}
        src="test-file-stub"
      />
      <div>
        <span
          className="info-icon-bg"
        />
        <span
          className="glyphicon glyphicon-info-sign"
          title="A new version (v1.2.4) of this app is available, and can be installed from the \\"Add/remove apps\\" screen"
        />
      </div>
    </div>
    <div>
      <h4
        className="list-group-item-heading"
      >
        pc-nrfconnect-foobar
      </h4>
      <p
        className="list-group-item-text"
      >
        , v
        1.2.3
      </p>
    </div>
    <div
      className="core-app-launch-item-buttons btn-toolbar"
    >
      <button
        className="btn btn-primary core-btn"
        disabled={false}
        onClick={[Function]}
        title="Launch app"
        type="button"
      >
        <span
          className="glyphicon glyphicon-play"
        />
        <span
          className="core-btn-text"
        >
          Launch
        </span>
      </button>
      <div
        className="dropdown btn-group core-dropdown"
      >
        <div
          className="dropdown btn-group"
        >
          <button
            aria-expanded={false}
            aria-haspopup={true}
            className="btn btn-primary core-btn dropdown-toggle btn btn-default"
            disabled={false}
            id="bg-nested-dropdown"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            type="button"
          >
             
            <span
              className="caret"
            />
          </button>
          <ul
            aria-labelledby="bg-nested-dropdown"
            className="dropdown-menu dropdown-menu-right"
            role="menu"
          >
            <li
              className=""
              role="presentation"
            >
              <a
                href="#"
                onClick={[Function]}
                onKeyDown={[Function]}
                role="menuitem"
                tabIndex="-1"
              >
                <button
                  className="btn btn-primary core-btn"
                  disabled={false}
                  onClick={[Function]}
                  title="Create a desktop shortcut for this app"
                  type="button"
                >
                  <span
                    className="glyphicon glyphicon-share-alt"
                  />
                  <span
                    className="core-btn-text"
                  >
                    Create shortcut
                  </span>
                </button>
              </a>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`AppLaunchView should render apps sorted by name and isOfficial 1`] = `
<div
  className="list-group"
>
  <div
    className="core-app-launch-item list-group-item"
  >
    <div
      className="core-app-icon"
    >
      <img
        alt="App icon"
        draggable={false}
        src="test-file-stub"
      />
    </div>
    <div>
      <h4
        className="list-group-item-heading"
      >
        appA
      </h4>
      <p
        className="list-group-item-text"
      >
        , v
        1.2.3
      </p>
    </div>
    <div
      className="core-app-launch-item-buttons btn-toolbar"
    >
      <button
        className="btn btn-primary core-btn"
        disabled={false}
        onClick={[Function]}
        title="Launch app"
        type="button"
      >
        <span
          className="glyphicon glyphicon-play"
        />
        <span
          className="core-btn-text"
        >
          Launch
        </span>
      </button>
      <div
        className="dropdown btn-group core-dropdown"
      >
        <div
          className="dropdown btn-group"
        >
          <button
            aria-expanded={false}
            aria-haspopup={true}
            className="btn btn-primary core-btn dropdown-toggle btn btn-default"
            disabled={false}
            id="bg-nested-dropdown"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            type="button"
          >
             
            <span
              className="caret"
            />
          </button>
          <ul
            aria-labelledby="bg-nested-dropdown"
            className="dropdown-menu dropdown-menu-right"
            role="menu"
          >
            <li
              className=""
              role="presentation"
            >
              <a
                href="#"
                onClick={[Function]}
                onKeyDown={[Function]}
                role="menuitem"
                tabIndex="-1"
              >
                <button
                  className="btn btn-primary core-btn"
                  disabled={false}
                  onClick={[Function]}
                  title="Create a desktop shortcut for this app"
                  type="button"
                >
                  <span
                    className="glyphicon glyphicon-share-alt"
                  />
                  <span
                    className="core-btn-text"
                  >
                    Create shortcut
                  </span>
                </button>
              </a>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>
  <div
    className="core-app-launch-item list-group-item"
  >
    <div
      className="core-app-icon"
    >
      <img
        alt="App icon"
        draggable={false}
        src="test-file-stub"
      />
    </div>
    <div>
      <h4
        className="list-group-item-heading"
      >
        appA
      </h4>
      <p
        className="list-group-item-text"
      >
        , v
        1.2.3
      </p>
    </div>
    <div
      className="core-app-launch-item-buttons btn-toolbar"
    >
      <button
        className="btn btn-primary core-btn"
        disabled={false}
        onClick={[Function]}
        title="Launch app"
        type="button"
      >
        <span
          className="glyphicon glyphicon-play"
        />
        <span
          className="core-btn-text"
        >
          Launch
        </span>
      </button>
      <div
        className="dropdown btn-group core-dropdown"
      >
        <div
          className="dropdown btn-group"
        >
          <button
            aria-expanded={false}
            aria-haspopup={true}
            className="btn btn-primary core-btn dropdown-toggle btn btn-default"
            disabled={false}
            id="bg-nested-dropdown"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            type="button"
          >
             
            <span
              className="caret"
            />
          </button>
          <ul
            aria-labelledby="bg-nested-dropdown"
            className="dropdown-menu dropdown-menu-right"
            role="menu"
          >
            <li
              className=""
              role="presentation"
            >
              <a
                href="#"
                onClick={[Function]}
                onKeyDown={[Function]}
                role="menuitem"
                tabIndex="-1"
              >
                <button
                  className="btn btn-primary core-btn"
                  disabled={false}
                  onClick={[Function]}
                  title="Create a desktop shortcut for this app"
                  type="button"
                >
                  <span
                    className="glyphicon glyphicon-share-alt"
                  />
                  <span
                    className="core-btn-text"
                  >
                    Create shortcut
                  </span>
                </button>
              </a>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>
  <div
    className="core-app-launch-item list-group-item"
  >
    <div
      className="core-app-icon"
    >
      <img
        alt="App icon"
        draggable={false}
        src="test-file-stub"
      />
    </div>
    <div>
      <h4
        className="list-group-item-heading"
      >
        appA
      </h4>
      <p
        className="list-group-item-text"
      >
        local
        , v
        1.2.3
      </p>
    </div>
    <div
      className="core-app-launch-item-buttons btn-toolbar"
    >
      <button
        className="btn btn-primary core-btn"
        disabled={false}
        onClick={[Function]}
        title="Launch app"
        type="button"
      >
        <span
          className="glyphicon glyphicon-play"
        />
        <span
          className="core-btn-text"
        >
          Launch
        </span>
      </button>
      <div
        className="dropdown btn-group core-dropdown"
      >
        <div
          className="dropdown btn-group"
        >
          <button
            aria-expanded={false}
            aria-haspopup={true}
            className="btn btn-primary core-btn dropdown-toggle btn btn-default"
            disabled={false}
            id="bg-nested-dropdown"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            type="button"
          >
             
            <span
              className="caret"
            />
          </button>
          <ul
            aria-labelledby="bg-nested-dropdown"
            className="dropdown-menu dropdown-menu-right"
            role="menu"
          >
            <li
              className=""
              role="presentation"
            >
              <a
                href="#"
                onClick={[Function]}
                onKeyDown={[Function]}
                role="menuitem"
                tabIndex="-1"
              >
                <button
                  className="btn btn-primary core-btn"
                  disabled={false}
                  onClick={[Function]}
                  title="Create a desktop shortcut for this app"
                  type="button"
                >
                  <span
                    className="glyphicon glyphicon-share-alt"
                  />
                  <span
                    className="core-btn-text"
                  >
                    Create shortcut
                  </span>
                </button>
              </a>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>
  <div
    className="core-app-launch-item list-group-item"
  >
    <div
      className="core-app-icon"
    >
      <img
        alt="App icon"
        draggable={false}
        src="test-file-stub"
      />
    </div>
    <div>
      <h4
        className="list-group-item-heading"
      >
        appB
      </h4>
      <p
        className="list-group-item-text"
      >
        , v
        1.2.3
      </p>
    </div>
    <div
      className="core-app-launch-item-buttons btn-toolbar"
    >
      <button
        className="btn btn-primary core-btn"
        disabled={false}
        onClick={[Function]}
        title="Launch app"
        type="button"
      >
        <span
          className="glyphicon glyphicon-play"
        />
        <span
          className="core-btn-text"
        >
          Launch
        </span>
      </button>
      <div
        className="dropdown btn-group core-dropdown"
      >
        <div
          className="dropdown btn-group"
        >
          <button
            aria-expanded={false}
            aria-haspopup={true}
            className="btn btn-primary core-btn dropdown-toggle btn btn-default"
            disabled={false}
            id="bg-nested-dropdown"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            type="button"
          >
             
            <span
              className="caret"
            />
          </button>
          <ul
            aria-labelledby="bg-nested-dropdown"
            className="dropdown-menu dropdown-menu-right"
            role="menu"
          >
            <li
              className=""
              role="presentation"
            >
              <a
                href="#"
                onClick={[Function]}
                onKeyDown={[Function]}
                role="menuitem"
                tabIndex="-1"
              >
                <button
                  className="btn btn-primary core-btn"
                  disabled={false}
                  onClick={[Function]}
                  title="Create a desktop shortcut for this app"
                  type="button"
                >
                  <span
                    className="glyphicon glyphicon-share-alt"
                  />
                  <span
                    className="core-btn-text"
                  >
                    Create shortcut
                  </span>
                </button>
              </a>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>
  <div
    className="core-app-launch-item list-group-item"
  >
    <div
      className="core-app-icon"
    >
      <img
        alt="App icon"
        draggable={false}
        src="test-file-stub"
      />
    </div>
    <div>
      <h4
        className="list-group-item-heading"
      >
        appB
      </h4>
      <p
        className="list-group-item-text"
      >
        local
        , v
        1.2.3
      </p>
    </div>
    <div
      className="core-app-launch-item-buttons btn-toolbar"
    >
      <button
        className="btn btn-primary core-btn"
        disabled={false}
        onClick={[Function]}
        title="Launch app"
        type="button"
      >
        <span
          className="glyphicon glyphicon-play"
        />
        <span
          className="core-btn-text"
        >
          Launch
        </span>
      </button>
      <div
        className="dropdown btn-group core-dropdown"
      >
        <div
          className="dropdown btn-group"
        >
          <button
            aria-expanded={false}
            aria-haspopup={true}
            className="btn btn-primary core-btn dropdown-toggle btn btn-default"
            disabled={false}
            id="bg-nested-dropdown"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            type="button"
          >
             
            <span
              className="caret"
            />
          </button>
          <ul
            aria-labelledby="bg-nested-dropdown"
            className="dropdown-menu dropdown-menu-right"
            role="menu"
          >
            <li
              className=""
              role="presentation"
            >
              <a
                href="#"
                onClick={[Function]}
                onKeyDown={[Function]}
                role="menuitem"
                tabIndex="-1"
              >
                <button
                  className="btn btn-primary core-btn"
                  disabled={false}
                  onClick={[Function]}
                  title="Create a desktop shortcut for this app"
                  type="button"
                >
                  <span
                    className="glyphicon glyphicon-share-alt"
                  />
                  <span
                    className="core-btn-text"
                  >
                    Create shortcut
                  </span>
                </button>
              </a>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`AppLaunchView should render spinner while retrieving apps 1`] = `
<div
  className="core-loading-apps-spinner"
/>
`;

exports[`AppLaunchView should render without any apps 1`] = `
<div
  className="list-group"
>
  <div>
    <h4>
      Welcome to nRF Connect
    </h4>
    <p>
      To get started, go to
      <i>
        Add/remove apps
      </i>
      to install some apps.
    </p>
  </div>
</div>
`;
